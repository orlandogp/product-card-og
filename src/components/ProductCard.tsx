import React, { CSSProperties, createContext } from 'react';
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/product';
import { useProduct } from '../hooks';
import styles from '../styles/styles.module.css';

export interface ProductCardProps {
  className?: string;
  initialValues?: InitialValues;
  product: Product;
  style?: CSSProperties;
  value?: number;
  children: (args: ProductCardHandlers) => JSX.Element;
  onChange?: (args: OnChangeArgs) => void;
}

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: ProductCardProps) => {
  const { counter, maxCount, isMaxCountReached, incBy, reset } = useProduct({
    initialValues,
    onChange,
    product,
    value,
  });

  return (
    <Provider value={{ counter, incBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.count,
          product,
          incBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
