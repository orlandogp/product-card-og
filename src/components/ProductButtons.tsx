import React, { CSSProperties, useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface ProductButtonsProps {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: ProductButtonsProps) => {
  const { counter, maxCount, incBy } = useContext(ProductContext);

  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
      <button
        type="button"
        className={styles.buttonMinus}
        onClick={() => incBy(-1)}
      >
        -
      </button>
      <div className={styles.countLabel}>{counter}</div>
      <button
        disabled={isMaxReached()}
        type="button"
        className={`${styles.buttonAdd} ${
          isMaxReached() ? styles.disabled : ''
        }`}
        onClick={() => incBy(1)}
      >
        +
      </button>
    </div>
  );
};
