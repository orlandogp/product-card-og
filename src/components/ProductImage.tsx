import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
import noImg from '../assets/no-image.jpg';

export interface ProductImageProps {
  className?: string;
  img?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductImage = ({
  img,
  title,
  className,
  style,
}: ProductImageProps) => {
  const { product } = useContext(ProductContext);

  const imgToShow = !img && !product.img ? noImg : img ? img : product.img;

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt={title}
      style={style}
    />
  );
};
