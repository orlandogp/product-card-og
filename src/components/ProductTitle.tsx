import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface ProductTitleProps {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({
  title,
  className,
  style,
}: ProductTitleProps) => {
  const { product } = useContext(ProductContext);
  const titleToShow = title ? title : product.title;

  return (
    <span className={`${styles.productDescription} ${className}`} style={style}>
      {titleToShow}
    </span>
  );
};
