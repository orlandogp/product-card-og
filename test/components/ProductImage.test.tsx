import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { productTwo } from '../data/products';

describe('ProductImage', () => {
  test('should show the component correctly with custom img', () => {
    const wrapper = renderer.create(<ProductImage img="https://hola.jpg" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('should show the component with product image', () => {
    const wrapper = renderer.create(
      <ProductCard product={productTwo}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
