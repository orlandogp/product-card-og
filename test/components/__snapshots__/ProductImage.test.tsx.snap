// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage should show the component correctly with custom img 1`] = `
<img
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;

exports[`ProductImage should show the component with product image 1`] = `
<div
  className="productCard undefined"
>
  <img
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
